File: CoreNet/Services/GenreService.cs
using CoreNet.Models;
using CoreNet.Repositories;

namespace CoreNet.Services
{
    public class GenreService
    {
        private readonly IGenreRepository _genreRepository;

        public GenreService(IGenreRepository genreRepository)
        {
            _genreRepository = genreRepository;
        }

        public Task<IEnumerable<Genre>> GetAllGenres() => _genreRepository.GetAllGenres();

        public Task<Genre?> GetGenreById(int id) => _genreRepository.GetGenreById(id);

        public Task CreateGenre(Genre genre) => _genreRepository.CreateGenre(genre);

        public async Task UpdateGenre(Genre genre)
        {
            var existingGenre = await _genreRepository.GetGenreById(genre.Id);
            if (existingGenre == null)
            {
                throw new Exception("Gênero não encontrado.");
            }

            await _genreRepository.UpdateGenre(genre);
        }

        public Task DeleteGenre(int id) => _genreRepository.DeleteGenre(id);
    }
}

File: node-api/src/services/genre.service.ts
import {
  Injectable,
  NotFoundException,
  InternalServerErrorException,
  BadRequestException,
} from '@nestjs/common'
import { GenreRepository } from '../repositories/genre.repository' 
import { CreateGenreDto } from '../models/create-genre.dto'
import { UpdateGenreDto } from '../models/update-genre.dto'

@Injectable()
export class GenreService {
  constructor(private readonly genreRepository: GenreRepository) {}

  async create(createGenreDto: CreateGenreDto) {
    try {
      return await this.genreRepository.createGenre(createGenreDto)
    } catch (error) {
      throw new InternalServerErrorException('Erro ao criar gênero de livro!')
    }
  }

  async findAll() {
    try {
      return await this.genreRepository.findAllGenres()
    } catch (error) {
      throw new InternalServerErrorException(
        'Erro ao buscar gêneros de livros!',
      )
    }
  }

  async findOne(id: string) {
    try {
      const genre = await this.genreRepository.findGenreById(id)
      if (!genre) {
        throw new NotFoundException('Gênero de livro não encontrado')
      }
      return genre
    } catch (error) {
      throw new InternalServerErrorException(
        'Erro ao buscar gênero de livro pelo ID!',
      )
    }
  }

  async update(id: string, updateGenreDto: UpdateGenreDto) {
    try {
      return await this.genreRepository.updateGenre(id, updateGenreDto)
    } catch (error) {
      throw new BadRequestException(
        'Dados inválidos para atualização do gênero de livro!',
      )
    }
  }

  async remove(id: string) {
    try {
      return await this.genreRepository.deleteGenre(id)
    } catch (error) {
      throw new NotFoundException(
        'Gênero de livro não encontrado para remoção!',
      )
    }
  }
}
